src/orange/detail/BIHTraverser.hh

```diff
@@ -0,0 +1,295 @@
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Point-in-volume operation.
+ */
+template<class F>
+CELER_FUNCTION LocalVolumeId BIHTraverser::operator()(Real3 const& point,
+                                                      F&& functor) const
+{
+    BIHNodeId previous_node;
+    BIHNodeId current_node{0};
+    LocalVolumeId id;
+
+    do
+    {
+        if (!is_inner(current_node))
+        {
+            id = visit_leaf(this->get_leaf_node(current_node), point, functor);
+
+            if (id)
+            {
+                return id;
+            }
+        }
+
+        auto temp = current_node;
+        current_node = this->next_node(current_node, previous_node, point);
+        previous_node = temp;
+
+    } while (current_node);
+
+    if (!id)
+    {
+        id = this->visit_inf_vols(functor);
+    }
+
+    return id;
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ *  Get the ID of the next node in the traversal sequence.
+ */
+CELER_FUNCTION
+BIHNodeId BIHTraverser::next_node(BIHNodeId const& current_id,
+                                  BIHNodeId const& previous_id,
+                                  Real3 const& point) const
+{
+    using Edge = BIHInnerNode::Edge;
+
+    BIHNodeId next_id;
+
+    if (is_inner(current_id))
+    {
+        auto const& current_node = this->get_inner_node(current_id);
+        if (previous_id == current_node.parent)
+        {
+            // Visiting this inner node for the first time; go down either left
+            // or right edge
+            if (this->visit_edge(current_node, Edge::left, point))
+            {
+                next_id = current_node.bounding_planes[Edge::left].child;
+            }
+            else
+            {
+                next_id = current_node.bounding_planes[Edge::right].child;
+            }
+        }
+        else if (previous_id == current_node.bounding_planes[Edge::left].child)
+        {
+            // Visiting this inner node for the second time; go down right edge
+            // or return to parent
+            if (this->visit_edge(current_node, Edge::right, point))
+            {
+                next_id = current_node.bounding_planes[Edge::right].child;
+            }
+            else
+            {
+                next_id = current_node.parent;
+            }
+        }
+        else
+        {
+            // Visiting this inner node for the third time; return to parent
+            CELER_EXPECT(previous_id
+                         == current_node.bounding_planes[Edge::right].child);
+            next_id = current_node.parent;
+        }
+    }
+    else
+    {
+        // Leaf node; return to parent
+        CELER_EXPECT(previous_id == this->get_leaf_node(current_id).parent);
+        next_id = previous_id;
+    }
+
+    return next_id;
+}
```

Comment:
This performance-critical code is rather branch heavy. Once we start doing optimization we should probably see if there's a cleaner way to do this.